### PR TITLE
Prioritize casper tx

### PR DESCRIPTION
**Pull request by kimmylin**
Prioritize Casper votes in the transactions of a block.

### CODE REVIEW

File: core/types/transaction.go [hunk 1]
```diff
@@ -360,9 +362,23 @@ func (s TxByNonce) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
 // for all at once sorting as well as individually adding and removing elements.
 type TxByPrice Transactions
 
-func (s TxByPrice) Len() int           { return len(s) }
-func (s TxByPrice) Less(i, j int) bool { return s[i].data.Price.Cmp(s[j].data.Price) > 0 }
-func (s TxByPrice) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+func (s TxByPrice) Len() int { return len(s) }
+
+func (s TxByPrice) Less(i, j int) bool {
+	// TODO: Check other tx fields to make sure it's a "vote"
```
Comment (changwu-tw) [inline core/types/transaction.go]: In theory, we can check the vote function by investigation the tx data to see whether the data start with particular hash.

However, we have several function which is valid in {send_valcode, login, vote, logout, withdraw}. Perhaps we can simply check the `to` field for now.

File: core/types/transaction.go [hunk 2]
```diff
@@ -35,6 +35,8 @@ import (
 var (
 	ErrInvalidSig = errors.New("invalid transaction v, r, s values")
 	errNoSigner   = errors.New("missing signing methods")
+	// TODO: Get Casper contract address from config?
```
Comment (changwu-tw) [inline core/types/transaction.go]: How about moving the casper address to a config file so other files can access it easily?

Comment (markya0616) [reply core/types/transaction.go]: How about put in `ChainConfig`?

File: core/types/transaction.go (hunk 1 above)
Comment (kimmylin) [reply core/types/transaction.go]: agreed. i think we should revisit this issue before we send this branch for review, hence the TODO

File: core/types/transaction.go (hunk 2 above)
Comment (kimmylin) [reply core/types/transaction.go]: i was going to wait for @markya0616 to finish the initialization stuff and we can make a final decision on where to stuff the address.

File: core/types/transaction.go [hunk 3]
```diff
@@ -360,9 +362,23 @@ func (s TxByNonce) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
...
-func (s TxByPrice) Less(i, j int) bool { return s[i].data.Price.Cmp(s[j].data.Price) > 0 }
-func (s TxByPrice) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+func (s TxByPrice) Len() int { return len(s) }
+
+func (s TxByPrice) Less(i, j int) bool {
+	// TODO: Check other tx fields to make sure it's a "vote"
+	if *s[i].data.Recipient == CasperAddr && *s[j].data.Recipient == CasperAddr {
+		return s[i].data.Price.Cmp(s[j].data.Price) > 0
+	}
+	if *s[i].data.Recipient == CasperAddr {
```
Comment (markya0616) [inline core/types/transaction.go]: `*s[i]` -> `s[i]`

File: core/types/transaction.go (hunk 1 above)
Comment (markya0616) [reply core/types/transaction.go]: Should we protect this by `g.Config.IsCasper(header.Number)`?

Comment (kimmylin) [reply core/types/transaction.go]: i'll have a separate PR to put the address in the config, and also guard the check by the config value.

**Pull request merged**